Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+      }
+    }
+    
+    mutating func send(status: String, trailingMetadata: Metadata) throws {
+      // Close the server.
+      switch self.state {
+      case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server can't send anything if idle.")
+      case .clientOpenServerOpen(let state):
+        self.state = .clientOpenServerClosed(.init(previousState: state))
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is closed, can't send anything else.")
+      case .clientClosedServerOpen(let state):
+        state.compressor?.end()
+        state.decompressor?.end()
+        self.state = .clientClosedServerClosed
+      case .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server can't send anything if closed.")
+      }
+    }
+
+    mutating func receive(metadata: Metadata, endStream: Bool) throws -> OnMetadataReceived {
+      if endStream {
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition(
+          """
+          Client should have opened before ending the stream:
+          stream shouldn't have been closed when sending initial metadata.
+          """
+        )
+      }
+      
+      guard let contentType = metadata.contentType else {
+        throw RPCError(code: .invalidArgument, message: "Invalid or empty content-type.")
+      }
+      
+      guard let endpoint = metadata.endpoint else {
+        throw RPCError(code: .unimplemented, message: "No :path header has been set.")
+      }
+      
+      // TODO: Should we verify the RPCRouter can handle this endpoint here,
+      // or should we verify that in the handler?
+      
+      let encodingValues = metadata[stringValues: "grpc-encoding"]
+      var encodingValuesIterator = encodingValues.makeIterator()
+      if let rawEncoding = encodingValuesIterator.next() {
+        guard encodingValuesIterator.next() == nil else {
+          throw RPCError(
+            code: .invalidArgument,
+            message: "grpc-encoding must contain no more than one value"
+          )
+        }
+        guard let encoding = Encoding(rawValue: rawEncoding) else {
+          let status = Status(
+            code: .unimplemented,
+            message: "\(rawEncoding) compression is not supported; supported algorithms are listed in grpc-accept-encoding"
+          )
+          let trailers = Metadata(dictionaryLiteral: (
+            "grpc-accept-encoding",
+            .string(self.supportedCompressionAlgorithms
+              .map({ $0.rawValue })
+              .joined(separator: ",")
+            )
+          ))
+          return .reject(status: status, trailers: trailers)
+        }
+
+        guard self.supportedCompressionAlgorithms.contains(where: { $0 == encoding }) else {
+          if self.supportedCompressionAlgorithms.isEmpty {
```

Comment:
We need to be careful that we always support `identity`, i.e. no compression. At the moment `encoding` could be `identity` and supported algorithms could be empty and we'd throw an error which would be wrong!